lib/membrane_mp4/cmaf/muxer.ex

```diff
@@ -71,7 +95,17 @@ defmodule Membrane.MP4.CMAF.Muxer do
         |> Header.serialize()
     }
 
-    {{:ok, caps: {:output, caps}, redemand: :output}, state}
+    # forwarding new caps action should be postponed so that
+    # discontinuity event arrives after the last buffer representing old
+    # caps is sent, cache the caps and handle propagating new caps only when handling new samples
+    state =
+      if caps != ctx.pads.output.caps do
+        %{state | pending_caps: caps}
```

Comment:
Since after the caps change the keyframe has to be sent immediately, we can generate the current segment here. That would simplify the code in handle_process.

Reply:
I guess it's not that simple as for generating a new segment we need a key_frame sample and we can't have that in`handle_caps`

Reply:
For generating a new segment we only need the frames that should be in that segment, not the next keyframe.

Reply:
The `generate_segment` function takes an argument `next_metadata` which in our case belongs to the next key  frames.
In handle_caps we have no additional sample that we can take the metadata from, that is why the caps are cached
and segment gets created in `handle_process`.

Reply:
Good point, haven't noticed. So at least we could avoid caching when `state.samples == []`

Reply:
Do we event need to cache the whole caps? In theory only a boolean flag should be enough because the caps should equal to the one in the context

Reply:
good point, the flag should be enough

Reply:
As discussed via DMs, the flow eventually needs to look like this:
* new caps arrive, cache the old caps and postpone forwarding of the new caps if there are cached segments representing old caps
* we reach `handle_process`, cache samples until we get a key frame
* if we reached key frame then create a segment using the old caps, forward the created segment as a buffer and right after that forward the current caps as they have been postponed
* remove the cached caps as they are no longer used